lightning/src/ln/channelmanager.rs

```diff
@@ -6189,6 +6220,255 @@ where
+											chan.context.is_usable() {
+											Some((chan, balances))
+										} else {
+											None
+										}
+									})
+									.min_by_key(|(_, balances)| balances.next_outbound_htlc_limit_msat).map(|(c, _)| c);
+								let optimal_channel = match maybe_optimal_channel {
+									Some(chan) => chan,
+									None => {
+										// Fall back to the specified channel to return an appropriate error.
+										if let Some(chan) = peer_state.channel_by_id
+											.get_mut(&forward_chan_id)
+											.and_then(Channel::as_funded_mut)
+										{
+											chan
+										} else {
+											push_trampoline_forwarding_failure(format!("Could not to route to next Trampoline hop {next_node_id} via forwarding channel {outgoing_scid}"), htlc_source, Some(outgoing_scid), 0x2000 | 25, Vec::new());
+											continue;
+										}
+									}
+								};
+
+								let logger = WithChannelContext::from(&self.logger, &optimal_channel.context, Some(payment_hash));
+								let channel_description = if optimal_channel.context.get_short_channel_id() == Some(short_chan_id) {
+									"specified"
+								} else {
+									"alternate"
+								};
+								log_trace!(logger, "Forwarding HTLC from SCID {} with payment_hash {} and next hop SCID {} over {} channel {} with corresponding peer {}",
+									prev_short_channel_id, &payment_hash, short_chan_id, channel_description, optimal_channel.context.channel_id(), &counterparty_node_id);
+								// Note that for inter-Trampoline forwards, we never add the blinding point to the UpdateAddHTLC message
+								if let Err(e) = optimal_channel.queue_add_htlc(outer_value_msat,
+									payment_hash, outer_cltv, htlc_source.clone(),
+									outer_onion_packet.clone(), skimmed_fee_msat, None, &self.fee_estimator,
+									&&logger)
+								{
+									if let ChannelError::Ignore(msg) = e {
+										log_trace!(logger, "Failed to forward HTLC with payment_hash {} to peer {}: {}", &payment_hash, &counterparty_node_id, msg);
+									} else {
+										panic!("Stated return value requirements in send_htlc() were not met");
+									}
+
+									if let Some(chan) = peer_state.channel_by_id
+										.get_mut(&forward_chan_id)
+										.and_then(Channel::as_funded_mut)
+									{
+										let failure_code = 0x1000|7;
+										let data = self.get_htlc_inbound_temp_fail_data(failure_code);
+										failed_forwards.push((htlc_source, payment_hash,
+											HTLCFailReason::reason(failure_code, data),
+											HTLCDestination::NextHopChannel { node_id: Some(chan.context.get_counterparty_node_id()), channel_id: forward_chan_id }
+										));
+									} else {
+										push_trampoline_forwarding_failure(format!("Could not to route to next Trampoline hop {next_node_id} via forwarding channel {outgoing_scid}"), htlc_source, Some(outgoing_scid), 0x2000 | 25, Vec::new());
+										continue;
+									}
+								}
+								()
+							},
```

Comment:
This is a lot of code added. Is it possible to do some decomposition in methods?

I would also suggest to add the maximum amount of comments in all of this code to explain as much as possible.